### PR TITLE
Improve docstrings for langchain.schema.py

### PR DESCRIPTION
**Pull request by dev2049**

### CODE REVIEW

File: langchain/schema.py [hunk 1]
```diff
@@ -51,58 +73,72 @@ def get_buffer_string(
 
 @dataclass
 class AgentAction:
-    """Agent's action to take."""
+    """A full description of an action for an ActionAgent to execute."""
 
     tool: str
+    """The name of the Tool to execute."""
     tool_input: Union[str, dict]
+    """The input to pass in to the Tool."""
```
Comment (hwchase17) [inline langchain/schema.py:81]: some more info here may be nice

File: langchain/schema.py [hunk 2]
```diff
@@ -51,58 +73,72 @@ def get_buffer_string(
...
     log: str
+    """Additional information to log about the action."""
```
Comment (hwchase17) [inline langchain/schema.py:83]: more info here may be nice

File: langchain/schema.py [hunk 3]
```diff
@@ -244,7 +307,7 @@ def flatten(self) -> List[LLMResult]:
                 )
             else:
                 if self.llm_output is not None:
-                    llm_output = self.llm_output.copy()
+                    llm_output = deepcopy(self.llm_output)
```
Comment (hwchase17) [inline langchain/schema.py:310]: this is intentional? why?

Comment (hwchase17) [reply langchain/schema.py:310]: test if so?

File: langchain/schema.py [hunk 4]
```diff
@@ -266,17 +330,50 @@ def __eq__(self, other: object) -> bool:
...
+        past executions of a Chain and inject that information into the inputs of
+        future executions of the Chain. For example, for conversational Chains Memory
+        can be used to store conversations and automatically add them to future model
+        prompts so that the model has the necessary context to respond coherently to
+        the latest input.
+
+    Example:
+        .. code-block:: python
+
+            class SimpleMemory(BaseMemory):
```
Comment (hwchase17) [inline langchain/schema.py:361]: does this class actually exist? 

I can see this being a bit confusing and I would maybe remove, im not sure this is conventional

File: langchain/schema.py [hunk 5]
```diff
@@ -353,21 +460,54 @@ def clear(self) -> None:
 
 
 class Document(Serializable):
-    """Interface for interacting with a document."""
+    """Class for storing a piece of text and associated metadata."""
 
     page_content: str
+    """String text."""
```
Comment (hwchase17) [inline langchain/schema.py:466]: lets beef up this description

File: langchain/schema.py [hunk 6]
```diff
@@ -353,21 +460,54 @@ def clear(self) -> None:
...
-    """Base interface for retrievers."""
+    """Abstract base class for a Document retrieval system.
+
+    A retrieval system is defined as something that can take string queries and return
+        the most 'relevant' Documents from some source.
+
+    Example:
+        .. code-block:: python
+
+            class TFIDFRetriever(BaseRetriever, BaseModel):
```
Comment (hwchase17) [inline langchain/schema.py:482]: again, not sure this is conventional and i can see this being confusing

File: langchain/schema.py [hunk 7]
```diff
@@ -378,63 +518,104 @@ async def aget_relevant_documents(self, query: str) -> List[Document]:
...
+        """
 
 
 class BaseOutputParser(BaseLLMOutputParser, ABC, Generic[T]):
     """Class to parse the output of an LLM call.
 
     Output parsers help structure language model responses.
-    """
+
+    Example:
```
Comment (hwchase17) [inline langchain/schema.py:557]: not sure this is convetional

File: langchain/schema.py [hunk 8]
```diff
@@ -378,63 +518,104 @@ async def aget_relevant_documents(self, query: str) -> List[Document]:
...
+            Structured output.
+        """
 
 
 class BaseOutputParser(BaseLLMOutputParser, ABC, Generic[T]):
     """Class to parse the output of an LLM call.
 
     Output parsers help structure language model responses.
-    """
+
```
Comment (hwchase17) [inline langchain/schema.py:556]: something about how this is differnt than BaseLLMOutputParser

File: langchain/schema.py [hunk 9]
```diff
@@ -501,16 +696,63 @@ def __init__(
 
 
 class BaseDocumentTransformer(ABC):
-    """Base interface for transforming documents."""
+    """Abstract base class for document transformation systems.
+
+    A document transformation system takes a sequence of Documents and returns a
+    sequence of transformed Documents.
+
+    Example:
```
Comment (hwchase17) [inline langchain/schema.py:704]: again not sure in convention

**Review by hwchase17 — COMMENTED** [review]
left a bunch of comments about examples of classes in the class docstrings. is this convetional? eg does sklearn do?

as im typing this, i can imagine this being really useful for classes we want people to write custom implementations for so maybe not the worst idea... maybe i take it back

**Comment by dev2049** [comment]
> left a bunch of comments about examples of classes in the class docstrings. is this convetional? eg does sklearn do?
> 
> as im typing this, i can imagine this being really useful for classes we want people to write custom implementations for so maybe not the worst idea... maybe i take it back

has similar thought process. there was one class that had an implementation in docstrings and i thought that was confusing (imagine examples are usually of usage). but then had thought that for abstract classes "usage" is implentation, so im now for it

File: langchain/schema.py (hunk 5 above)
Comment (dev2049) [reply langchain/schema.py:466]: not really sure what more to put given how generic idea of a Document is

File: langchain/schema.py (hunk 3 above)
Comment (dev2049) [reply langchain/schema.py:310]: couldn't think of a good reason we'd want to potentially share mutable objects across the flattened generations, assumed was mistake to not deep copy in first place. are you saying it's intentional to do shallow copy? don't love sharing dicts across objects, find it confusing

File: langchain/schema.py (hunk 8 above)
Comment (dev2049) [reply langchain/schema.py:556]: tbh do we want most folks to look at BaseLLMOutputParser? kinda feels like this is the much more generic thing and maybe we somehow hide the BaseLLMOutputParser more

File: langchain/schema.py [hunk 10]
```diff
@@ -51,58 +73,72 @@ def get_buffer_string(
...
     @property
     def lc_serializable(self) -> bool:
-        """This class is LangChain serializable."""
+        """Whether this class is LangChain serializable."""
         return True
 
 
 class BaseMessage(Serializable):
-    """Message object."""
+    """The base abstract Message class. Messages are the inputs and outputs of
```
[the remark this answers is not in the document]
Comment (hwchase17) [reply langchain/schema.py:114]: +1

**Review by hwchase17 — APPROVED** [review]

**Pull request merged**